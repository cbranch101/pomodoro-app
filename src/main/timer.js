const getDeferred = () => {
    const deferred = {}
    const promise = new Promise((resolve, reject) => {
        deferred.resolve = resolve
        deferred.reject = reject
    })
    return Object.assign({}, deferred, { promise })
}

const getTimerHandler = onTick => {
    const timers = {}
    const stop = (timerName, isCompleted = false) => {
        const currentTimer = timers[timerName]
        if (currentTimer) {
            const { handler, deferred, count } = currentTimer
            clearInterval(handler)
            deferred.resolve({ count, isCompleted })
            delete timers[timerName]
        }
    }
    const getTick = (timerName, stopFunc, getReturnValue) => () => {
        const timer = timers[timerName]
        const { count } = timer
        const newCount = count + 1
        timers[timerName].count = newCount
        const returnValue = getReturnValue ? getReturnValue(newCount) : newCount
        onTick(timerName, returnValue)
        if (stopFunc && stopFunc(newCount)) {
            stop(timerName, true)
        }
    }
    const start = (timerName, stopFunc, getReturnValue) => {
        const timer = setInterval(getTick(timerName, stopFunc, getReturnValue), 1000)
        const deferred = getDeferred()
        timers[timerName] = {
            deferred,
            handler: timer,
            count: 0
        }
        return deferred.promise
    }
    return {
        startFor: (timerName, duration) =>
            start(
                timerName,
                count => count === duration,
                count => ({
                    current: count,
                    remaining: duration - count
                })
            ),
        start,
        stop
    }
}

module.exports = {
    getTimerHandler
}
