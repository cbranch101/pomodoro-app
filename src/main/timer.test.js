import { getTimerHandler } from "./timer.js"

describe("timer", () => {
    test("the on tick handler should be called on every tick", done => {
        const timerHandler = getTimerHandler((timerName, count) => {
            expect(timerName).toEqual("test-timer")
            expect(count).toEqual(1)
            done()
        })
        timerHandler.startFor("test-timer", 1)
    })
    test(
        "all timer should be stoppable, and it should " +
            "be possible to have multiple timers running at once",
        async () => {
            const timerHandler = getTimerHandler(() => {})
            setTimeout(() => {
                timerHandler.stop("stop")
            }, 1050)
            const returnValue = await Promise.all([
                timerHandler.startFor("stop", 2),
                timerHandler.startFor("finish", 2)
            ])
            expect(returnValue).toMatchSnapshot()
        }
    )
})
