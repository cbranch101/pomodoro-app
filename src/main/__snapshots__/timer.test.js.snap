exports[`timer all timer should be stoppable, and it should be possible to have multiple timers running at once 1`] = `
Array [
  Object {
    "count": 1,
    "isCompleted": false,
  },
  Object {
    "count": 2,
    "isCompleted": true,
  },
]
`;
